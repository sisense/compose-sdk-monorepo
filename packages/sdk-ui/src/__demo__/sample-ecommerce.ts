import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
