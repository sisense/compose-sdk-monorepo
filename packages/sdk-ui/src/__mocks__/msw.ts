import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import * as mockGlobals from '@/__mocks__/data/mock-globals.json';
import * as mockPalettes from '@/__mocks__/data/mock-palettes.json';

export const mockUrl = 'http://fake-url/';
export const mockToken = 'fake-token';
export const mockDashboardId = 'fake-dashboard-id';
export const mockWidgetId = 'fake-widget-id';

const handlers = [
  // Mock common calls made by SisenseContextProvider
  http.get('*/api/globals', () => HttpResponse.json(mockGlobals)),
  http.get('*/api/palettes/*', () => HttpResponse.json(mockPalettes)),

  // Mock tracking calls
  http.post('*/api/activities', () => new HttpResponse()),
];

export const server = setupServer(...handlers);
