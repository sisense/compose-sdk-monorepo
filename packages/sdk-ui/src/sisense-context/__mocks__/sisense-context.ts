export const useSisenseContext = vi.fn();
export const useSisenseContextMock = useSisenseContext;
