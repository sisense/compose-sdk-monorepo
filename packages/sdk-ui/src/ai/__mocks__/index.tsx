import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import { mockToken, mockUrl } from '@/__mocks__/msw';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import AiContextProvider from '../ai-context-provider';

/**
 * A component that wraps an AI hook/component with all required providers.
 * For use in unit tests only.
 *
 * @internal
 */
export const AiTestWrapper = ({ children }: { children: ReactNode }) => (
  <SisenseContextProvider url={mockUrl} token={mockToken}>
    <AiContextProvider>
      <QueryClientProvider
        client={
          new QueryClient({
            defaultOptions: {
              queries: {
                retry: false,
              },
            },
          })
        }
      >
        {children}
      </QueryClientProvider>
    </AiContextProvider>
  </SisenseContextProvider>
);
