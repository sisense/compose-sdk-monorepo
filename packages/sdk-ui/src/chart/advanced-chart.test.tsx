import { render, waitFor } from '@testing-library/react';
import { http, HttpResponse } from 'msw';
import { measureFactory } from '@ethings-os/sdk-data';
import { Chart, HighchartsOptions, SisenseContextProviderProps } from '@/index';
import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import * as jaqlForecast from '@/__mocks__/data/mock-jaql-forecast.json';
import * as jaqlTrend from '@/__mocks__/data/mock-jaql-trend.json';
import * as DM from '@/__test-helpers__/sample-ecommerce';
import { mockToken, mockUrl, server } from '@/__mocks__/msw';

const contextProviderProps: SisenseContextProviderProps = {
  url: mockUrl,
  token: mockToken,
  defaultDataSource: 'Sample ECommerce',
  appConfig: {
    queryCacheConfig: { enabled: false },
    trackingConfig: { enabled: false },
  },
};

describe('Advanced Charts', () => {
  beforeEach(() => {
    server.resetHandlers();
  });
  it('should render forecast trend combo chart', async () => {
    server.use(
      http.post('*/api/datasources/:dataSource/jaql', () => HttpResponse.json(jaqlForecast)),
    );

    let preparedHighchartsOptions: HighchartsOptions | undefined;

    const { findByLabelText } = render(
      <SisenseContextProvider {...contextProviderProps}>
        <Chart
          dataSet={DM.DataSource}
          chartType={'line'}
          dataOptions={{
            category: [DM.Commerce.Date.Months],
            value: [
              {
                column: measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue'),
                forecast: {
                  forecastHorizon: 6,
                },
              },
              {
                column: measureFactory.sum(DM.Commerce.Cost, 'Total Cost'),
                forecast: {
                  forecastHorizon: 6,
                },
                trend: {
                  modelType: 'advancedSmoothing',
                },
              },
            ],
            breakBy: [],
          }}
          styleOptions={{ height: 600 }}
          onBeforeRender={(options: HighchartsOptions) => {
            preparedHighchartsOptions = options;
            return options;
          }}
        />
      </SisenseContextProvider>,
    );

    await waitFor(() => {
      expect(preparedHighchartsOptions).toBeDefined();
      expect(preparedHighchartsOptions).toMatchSnapshot();
    });

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });

  it('should render trend chart', async () => {
    server.use(http.post('*/api/datasources/:dataSource/jaql', () => HttpResponse.json(jaqlTrend)));

    let preparedHighchartsOptions: HighchartsOptions | undefined;

    const { findByLabelText } = render(
      <SisenseContextProvider {...contextProviderProps}>
        <Chart
          dataSet={DM.DataSource}
          chartType={'line'}
          dataOptions={{
            category: [DM.Commerce.Date.Months],
            value: [
              {
                column: measureFactory.sum(DM.Commerce.Cost, 'Cost'),
                trend: {
                  modelType: 'advancedSmoothing',
                },
              },
            ],
            breakBy: [],
          }}
          onBeforeRender={(options: HighchartsOptions) => {
            preparedHighchartsOptions = options;
            return options;
          }}
        />
      </SisenseContextProvider>,
    );

    await waitFor(() => {
      expect(preparedHighchartsOptions).toBeDefined();
      expect(preparedHighchartsOptions).toMatchSnapshot();
    });

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });
});
