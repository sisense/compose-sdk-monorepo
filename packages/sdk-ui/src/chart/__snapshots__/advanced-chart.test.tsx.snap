// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Advanced Charts > should render trend chart 1`] = `
{
  "accessibility": {
    "enabled": false,
  },
  "boost": {
    "useGPUTranslations": true,
    "usePreAllocated": true,
  },
  "chart": {
    "alignTicks": false,
    "animation": {
      "duration": 300,
    },
    "backgroundColor": "#FFFFFF",
    "events": {},
    "polar": false,
    "spacing": [
      20,
      20,
      20,
      20,
    ],
    "type": "line",
  },
  "credits": {
    "enabled": false,
  },
  "drilldown": {
    "activeDataLabelStyle": {
      "color": "#5B6372",
    },
    "breadcrumbs": {
      "buttonTheme": {
        "fill": "#ffcb05",
        "states": {
          "hover": {
            "fill": "#f2b900",
            "stroke": "#f2b900",
          },
        },
        "stroke": "#ffcb05",
        "style": {
          "color": "#3a4356",
        },
      },
    },
  },
  "exporting": {
    "enabled": true,
  },
  "legend": {
    "align": "center",
    "backgroundColor": "transparent",
    "borderColor": "transparent",
    "borderWidth": 0,
    "enabled": true,
    "itemStyle": {
      "color": "#5B6372",
      "cursor": "default",
      "fontFamily": "Open Sans",
      "fontSize": "13px",
      "fontWeight": "normal",
      "pointerEvents": "auto",
      "textOutline": "none",
    },
    "layout": "horizontal",
    "symbolRadius": 0,
    "verticalAlign": "bottom",
  },
  "navigator": {
    "enabled": true,
    "handles": {
      "backgroundColor": "#CCCCCC",
      "borderColor": "#FFFFFF",
      "symbols": [
        "navigator-handle",
        "navigator-handle",
      ],
    },
    "height": 40,
    "margin": 30,
    "maskFill": "rgba(91,99,114, 0.15)",
    "maskInside": true,
    "outlineColor": "#CCCCCC",
    "outlineWidth": 1,
    "series": {
      "dataGrouping": {
        "enabled": false,
      },
      "type": "line",
    },
    "tooltipFormatter": [Function],
    "xAxis": {
      "labels": {
        "enabled": false,
      },
    },
  },
  "plotOptions": {
    "area": {
      "className": "area-series--without-pointer-events",
      "states": {
        "inactive": {
          "enabled": true,
          "opacity": 0.3,
        },
      },
    },
    "series": {
      "animation": {
        "duration": 600,
      },
      "connectNulls": false,
      "dataLabels": {
        "crop": false,
        "enabled": false,
        "style": {
          "color": "#5B6372",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "marker": {
        "enabled": false,
        "fillColor": null,
        "fillOpacity": 0.3,
        "lineColor": null,
        "lineWidth": 2,
        "radius": 4,
        "states": {
          "hover": {
            "fillColor": "#FFFFFF",
            "lineWidth": 3,
            "radius": 6,
          },
        },
        "symbol": "circle",
      },
      "point": {
        "events": {
          "click": [Function],
        },
      },
      "stacking": undefined,
      "states": {
        "select": {
          "borderColor": "transparent",
          "color": null,
          "opacity": 0.3,
        },
      },
      "stickyTracking": false,
    },
  },
  "series": [
    {
      "boostThreshold": 0,
      "color": "#00cee6",
      "connectNulls": false,
      "data": [
        {
          "custom": {
            "rawValue": 189976.80757868662,
            "xValue": [
              "2009-11-01",
            ],
          },
          "selected": false,
          "y": 189976.80757868662,
        },
        {
          "custom": {
            "rawValue": 1107490.0728950314,
            "xValue": [
              "2009-12-01",
            ],
          },
          "selected": false,
          "y": 1107490.0728950314,
        },
        {
          "custom": {
            "rawValue": 845205.6287006252,
            "xValue": [
              "2010-01-01",
            ],
          },
          "selected": false,
          "y": 845205.6287006252,
        },
        {
          "custom": {
            "rawValue": 787917.8207888827,
            "xValue": [
              "2010-02-01",
            ],
          },
          "selected": false,
          "y": 787917.8207888827,
        },
        {
          "custom": {
            "rawValue": 946810.4692947622,
            "xValue": [
              "2010-03-01",
            ],
          },
          "selected": false,
          "y": 946810.4692947622,
        },
        {
          "custom": {
            "rawValue": 900508.5092079439,
            "xValue": [
              "2010-04-01",
            ],
          },
          "selected": false,
          "y": 900508.5092079439,
        },
        {
          "custom": {
            "rawValue": 946238.5815005917,
            "xValue": [
              "2010-05-01",
            ],
          },
          "selected": false,
          "y": 946238.5815005917,
        },
        {
          "custom": {
            "rawValue": 886700.2053317726,
            "xValue": [
              "2010-06-01",
            ],
          },
          "selected": false,
          "y": 886700.2053317726,
        },
        {
          "custom": {
            "rawValue": 900954.317531623,
            "xValue": [
              "2010-07-01",
            ],
          },
          "selected": false,
          "y": 900954.317531623,
        },
        {
          "custom": {
            "rawValue": 984546.9012292922,
            "xValue": [
              "2010-08-01",
            ],
          },
          "selected": false,
          "y": 984546.9012292922,
        },
        {
          "custom": {
            "rawValue": 947657.5594753642,
            "xValue": [
              "2010-09-01",
            ],
          },
          "selected": false,
          "y": 947657.5594753642,
        },
        {
          "custom": {
            "rawValue": 1067067.8516010046,
            "xValue": [
              "2010-10-01",
            ],
          },
          "selected": false,
          "y": 1067067.8516010046,
        },
        {
          "custom": {
            "rawValue": 1313277.0609966572,
            "xValue": [
              "2010-11-01",
            ],
          },
          "selected": false,
          "y": 1313277.0609966572,
        },
        {
          "custom": {
            "rawValue": 1467710.2702174708,
            "xValue": [
              "2010-12-01",
            ],
          },
          "selected": false,
          "y": 1467710.2702174708,
        },
        {
          "custom": {
            "rawValue": 1175238.5499559864,
            "xValue": [
              "2011-01-01",
            ],
          },
          "selected": false,
          "y": 1175238.5499559864,
        },
        {
          "custom": {
            "rawValue": 1092969.080261767,
            "xValue": [
              "2011-02-01",
            ],
          },
          "selected": false,
          "y": 1092969.080261767,
        },
        {
          "custom": {
            "rawValue": 1188733.9370134315,
            "xValue": [
              "2011-03-01",
            ],
          },
          "selected": false,
          "y": 1188733.9370134315,
        },
        {
          "custom": {
            "rawValue": 1468171.674996743,
            "xValue": [
              "2011-04-01",
            ],
          },
          "selected": false,
          "y": 1468171.674996743,
        },
        {
          "custom": {
            "rawValue": 1809045.4618380964,
            "xValue": [
              "2011-05-01",
            ],
          },
          "selected": false,
          "y": 1809045.4618380964,
        },
        {
          "custom": {
            "rawValue": 1963988.3454919308,
            "xValue": [
              "2011-06-01",
            ],
          },
          "selected": false,
          "y": 1963988.3454919308,
        },
        {
          "custom": {
            "rawValue": 2011346.4261528384,
            "xValue": [
              "2011-07-01",
            ],
          },
          "selected": false,
          "y": 2011346.4261528384,
        },
        {
          "custom": {
            "rawValue": 2128043.9337140303,
            "xValue": [
              "2011-08-01",
            ],
          },
          "selected": false,
          "y": 2128043.9337140303,
        },
        {
          "custom": {
            "rawValue": 2017077.2452351712,
            "xValue": [
              "2011-09-01",
            ],
          },
          "selected": false,
          "y": 2017077.2452351712,
        },
        {
          "custom": {
            "rawValue": 2283585.5977138802,
            "xValue": [
              "2011-10-01",
            ],
          },
          "selected": false,
          "y": 2283585.5977138802,
        },
        {
          "custom": {
            "rawValue": 2558169.054697007,
            "xValue": [
              "2011-11-01",
            ],
          },
          "selected": false,
          "y": 2558169.054697007,
        },
        {
          "custom": {
            "rawValue": 2855504.8063853886,
            "xValue": [
              "2011-12-01",
            ],
          },
          "selected": false,
          "y": 2855504.8063853886,
        },
        {
          "custom": {
            "rawValue": 2267886.0520572066,
            "xValue": [
              "2012-01-01",
            ],
          },
          "selected": false,
          "y": 2267886.0520572066,
        },
        {
          "custom": {
            "rawValue": 2289422.9980101623,
            "xValue": [
              "2012-02-01",
            ],
          },
          "selected": false,
          "y": 2289422.9980101623,
        },
        {
          "custom": {
            "rawValue": 2413584.21836257,
            "xValue": [
              "2012-03-01",
            ],
          },
          "selected": false,
          "y": 2413584.21836257,
        },
        {
          "custom": {
            "rawValue": 2450628.8393439148,
            "xValue": [
              "2012-04-01",
            ],
          },
          "selected": false,
          "y": 2450628.8393439148,
        },
        {
          "custom": {
            "rawValue": 2616360.1553670047,
            "xValue": [
              "2012-05-01",
            ],
          },
          "selected": false,
          "y": 2616360.1553670047,
        },
        {
          "custom": {
            "rawValue": 2482130.5598226488,
            "xValue": [
              "2012-06-01",
            ],
          },
          "selected": false,
          "y": 2482130.5598226488,
        },
        {
          "custom": {
            "rawValue": 2594841.3182018436,
            "xValue": [
              "2012-07-01",
            ],
          },
          "selected": false,
          "y": 2594841.3182018436,
        },
        {
          "custom": {
            "rawValue": 2673938.1166581884,
            "xValue": [
              "2012-08-01",
            ],
          },
          "selected": false,
          "y": 2673938.1166581884,
        },
        {
          "custom": {
            "rawValue": 2867416.29551314,
            "xValue": [
              "2012-09-01",
            ],
          },
          "selected": false,
          "y": 2867416.29551314,
        },
        {
          "custom": {
            "rawValue": 3168871.066950053,
            "xValue": [
              "2012-10-01",
            ],
          },
          "selected": false,
          "y": 3168871.066950053,
        },
        {
          "custom": {
            "rawValue": 3888213.687782809,
            "xValue": [
              "2012-11-01",
            ],
          },
          "selected": false,
          "y": 3888213.687782809,
        },
        {
          "custom": {
            "rawValue": 3992612.938552344,
            "xValue": [
              "2012-12-01",
            ],
          },
          "selected": false,
          "y": 3992612.938552344,
        },
        {
          "custom": {
            "rawValue": 2759434.9833942996,
            "xValue": [
              "2013-01-01",
            ],
          },
          "selected": false,
          "y": 2759434.9833942996,
        },
        {
          "custom": {
            "rawValue": 2788158.3545121085,
            "xValue": [
              "2013-02-01",
            ],
          },
          "selected": false,
          "y": 2788158.3545121085,
        },
        {
          "custom": {
            "rawValue": 3046625.3691154122,
            "xValue": [
              "2013-03-01",
            ],
          },
          "selected": false,
          "y": 3046625.3691154122,
        },
        {
          "custom": {
            "rawValue": 2961134.7611477133,
            "xValue": [
              "2013-04-01",
            ],
          },
          "selected": false,
          "y": 2961134.7611477133,
        },
        {
          "custom": {
            "rawValue": 3242762.429729399,
            "xValue": [
              "2013-05-01",
            ],
          },
          "selected": false,
          "y": 3242762.429729399,
        },
        {
          "custom": {
            "rawValue": 3204714.5343812816,
            "xValue": [
              "2013-06-01",
            ],
          },
          "selected": false,
          "y": 3204714.5343812816,
        },
        {
          "custom": {
            "rawValue": 3349081.2784886193,
            "xValue": [
              "2013-07-01",
            ],
          },
          "selected": false,
          "y": 3349081.2784886193,
        },
        {
          "custom": {
            "rawValue": 3238694.8539762036,
            "xValue": [
              "2013-08-01",
            ],
          },
          "selected": false,
          "y": 3238694.8539762036,
        },
        {
          "custom": {
            "rawValue": 3581952.4929154813,
            "xValue": [
              "2013-09-01",
            ],
          },
          "selected": false,
          "y": 3581952.4929154813,
        },
        {
          "custom": {
            "rawValue": 3935084.4678387493,
            "xValue": [
              "2013-10-01",
            ],
          },
          "selected": false,
          "y": 3935084.4678387493,
        },
        {
          "custom": {
            "rawValue": 4844611.66696444,
            "xValue": [
              "2013-11-01",
            ],
          },
          "selected": false,
          "y": 4844611.66696444,
        },
        {
          "custom": {
            "rawValue": 770441.3868424296,
            "xValue": [
              "2013-12-01",
            ],
          },
          "selected": false,
          "y": 770441.3868424296,
        },
      ],
      "dataLabels": {
        "formatter": [Function],
      },
      "lineWidth": 3,
      "marker": {
        "enabled": false,
        "fillColor": null,
        "fillOpacity": 0.3,
        "lineColor": null,
        "lineWidth": 2,
        "radius": 4,
        "states": {
          "hover": {
            "fillColor": "#FFFFFF",
            "lineWidth": 3,
            "radius": 6,
          },
        },
        "symbol": "circle",
      },
      "name": "Cost (+Trend)",
      "showInNavigator": true,
      "stickyTracking": false,
      "turboThreshold": 0,
      "yAxis": 0,
    },
    {
      "boostThreshold": 0,
      "color": "#00cee6",
      "connectNulls": false,
      "dashStyle": "ShortDot",
      "data": [
        {
          "custom": {
            "rawValue": 399877.66021296044,
            "xValue": [
              "2009-11-01",
            ],
          },
          "selected": false,
          "y": 399877.66021296044,
        },
        {
          "custom": {
            "rawValue": 760254.7486363016,
            "xValue": [
              "2009-12-01",
            ],
          },
          "selected": false,
          "y": 760254.7486363016,
        },
        {
          "custom": {
            "rawValue": 892589.3891729144,
            "xValue": [
              "2010-01-01",
            ],
          },
          "selected": false,
          "y": 892589.3891729144,
        },
        {
          "custom": {
            "rawValue": 904228.5180767283,
            "xValue": [
              "2010-02-01",
            ],
          },
          "selected": false,
          "y": 904228.5180767283,
        },
        {
          "custom": {
            "rawValue": 912866.4678721044,
            "xValue": [
              "2010-03-01",
            ],
          },
          "selected": false,
          "y": 912866.4678721044,
        },
        {
          "custom": {
            "rawValue": 916526.9340331501,
            "xValue": [
              "2010-04-01",
            ],
          },
          "selected": false,
          "y": 916526.9340331501,
        },
        {
          "custom": {
            "rawValue": 914095.3329283679,
            "xValue": [
              "2010-05-01",
            ],
          },
          "selected": false,
          "y": 914095.3329283679,
        },
        {
          "custom": {
            "rawValue": 904960.4685997746,
            "xValue": [
              "2010-06-01",
            ],
          },
          "selected": false,
          "y": 904960.4685997746,
        },
        {
          "custom": {
            "rawValue": 912084.7216383086,
            "xValue": [
              "2010-07-01",
            ],
          },
          "selected": false,
          "y": 912084.7216383086,
        },
        {
          "custom": {
            "rawValue": 945423.4204832909,
            "xValue": [
              "2010-08-01",
            ],
          },
          "selected": false,
          "y": 945423.4204832909,
        },
        {
          "custom": {
            "rawValue": 1009017.2938368567,
            "xValue": [
              "2010-09-01",
            ],
          },
          "selected": false,
          "y": 1009017.2938368567,
        },
        {
          "custom": {
            "rawValue": 1127908.0165300977,
            "xValue": [
              "2010-10-01",
            ],
          },
          "selected": false,
          "y": 1127908.0165300977,
        },
        {
          "custom": {
            "rawValue": 1260348.35798684,
            "xValue": [
              "2010-11-01",
            ],
          },
          "selected": false,
          "y": 1260348.35798684,
        },
        {
          "custom": {
            "rawValue": 1301643.164863898,
            "xValue": [
              "2010-12-01",
            ],
          },
          "selected": false,
          "y": 1301643.164863898,
        },
        {
          "custom": {
            "rawValue": 1226332.721934023,
            "xValue": [
              "2011-01-01",
            ],
          },
          "selected": false,
          "y": 1226332.721934023,
        },
        {
          "custom": {
            "rawValue": 1175483.052081485,
            "xValue": [
              "2011-02-01",
            ],
          },
          "selected": false,
          "y": 1175483.052081485,
        },
        {
          "custom": {
            "rawValue": 1259954.2895959807,
            "xValue": [
              "2011-03-01",
            ],
          },
          "selected": false,
          "y": 1259954.2895959807,
        },
        {
          "custom": {
            "rawValue": 1477513.9658073424,
            "xValue": [
              "2011-04-01",
            ],
          },
          "selected": false,
          "y": 1477513.9658073424,
        },
        {
          "custom": {
            "rawValue": 1731745.1938052739,
            "xValue": [
              "2011-05-01",
            ],
          },
          "selected": false,
          "y": 1731745.1938052739,
        },
        {
          "custom": {
            "rawValue": 1918408.774731447,
            "xValue": [
              "2011-06-01",
            ],
          },
          "selected": false,
          "y": 1918408.774731447,
        },
        {
          "custom": {
            "rawValue": 2019625.0977478286,
            "xValue": [
              "2011-07-01",
            ],
          },
          "selected": false,
          "y": 2019625.0977478286,
        },
        {
          "custom": {
            "rawValue": 2080349.5218863264,
            "xValue": [
              "2011-08-01",
            ],
          },
          "selected": false,
          "y": 2080349.5218863264,
        },
        {
          "custom": {
            "rawValue": 2160577.230230905,
            "xValue": [
              "2011-09-01",
            ],
          },
          "selected": false,
          "y": 2160577.230230905,
        },
        {
          "custom": {
            "rawValue": 2330053.153545553,
            "xValue": [
              "2011-10-01",
            ],
          },
          "selected": false,
          "y": 2330053.153545553,
        },
        {
          "custom": {
            "rawValue": 2515328.951597531,
            "xValue": [
              "2011-11-01",
            ],
          },
          "selected": false,
          "y": 2515328.951597531,
        },
        {
          "custom": {
            "rawValue": 2571323.757848365,
            "xValue": [
              "2011-12-01",
            ],
          },
          "selected": false,
          "y": 2571323.757848365,
        },
        {
          "custom": {
            "rawValue": 2455225.182035394,
            "xValue": [
              "2012-01-01",
            ],
          },
          "selected": false,
          "y": 2455225.182035394,
        },
        {
          "custom": {
            "rawValue": 2372277.1257133014,
            "xValue": [
              "2012-02-01",
            ],
          },
          "selected": false,
          "y": 2372277.1257133014,
        },
        {
          "custom": {
            "rawValue": 2390716.0906670545,
            "xValue": [
              "2012-03-01",
            ],
          },
          "selected": false,
          "y": 2390716.0906670545,
        },
        {
          "custom": {
            "rawValue": 2454806.7004078142,
            "xValue": [
              "2012-04-01",
            ],
          },
          "selected": false,
          "y": 2454806.7004078142,
        },
        {
          "custom": {
            "rawValue": 2509923.7089504977,
            "xValue": [
              "2012-05-01",
            ],
          },
          "selected": false,
          "y": 2509923.7089504977,
        },
        {
          "custom": {
            "rawValue": 2529590.1527741267,
            "xValue": [
              "2012-06-01",
            ],
          },
          "selected": false,
          "y": 2529590.1527741267,
        },
        {
          "custom": {
            "rawValue": 2580856.1512703863,
            "xValue": [
              "2012-07-01",
            ],
          },
          "selected": false,
          "y": 2580856.1512703863,
        },
        {
          "custom": {
            "rawValue": 2713417.6342164744,
            "xValue": [
              "2012-08-01",
            ],
          },
          "selected": false,
          "y": 2713417.6342164744,
        },
        {
          "custom": {
            "rawValue": 2969220.9206936043,
            "xValue": [
              "2012-09-01",
            ],
          },
          "selected": false,
          "y": 2969220.9206936043,
        },
        {
          "custom": {
            "rawValue": 3328777.947662453,
            "xValue": [
              "2012-10-01",
            ],
          },
          "selected": false,
          "y": 3328777.947662453,
        },
        {
          "custom": {
            "rawValue": 3620949.4273355627,
            "xValue": [
              "2012-11-01",
            ],
          },
          "selected": false,
          "y": 3620949.4273355627,
        },
        {
          "custom": {
            "rawValue": 3556054.100029769,
            "xValue": [
              "2012-12-01",
            ],
          },
          "selected": false,
          "y": 3556054.100029769,
        },
        {
          "custom": {
            "rawValue": 3180837.955961699,
            "xValue": [
              "2013-01-01",
            ],
          },
          "selected": false,
          "y": 3180837.955961699,
        },
        {
          "custom": {
            "rawValue": 2940071.145840517,
            "xValue": [
              "2013-02-01",
            ],
          },
          "selected": false,
          "y": 2940071.145840517,
        },
        {
          "custom": {
            "rawValue": 2928282.3596065277,
            "xValue": [
              "2013-03-01",
            ],
          },
          "selected": false,
          "y": 2928282.3596065277,
        },
        {
          "custom": {
            "rawValue": 3012322.5051070023,
            "xValue": [
              "2013-04-01",
            ],
          },
          "selected": false,
          "y": 3012322.5051070023,
        },
        {
          "custom": {
            "rawValue": 3126610.365876169,
            "xValue": [
              "2013-05-01",
            ],
          },
          "selected": false,
          "y": 3126610.365876169,
        },
        {
          "custom": {
            "rawValue": 3213000.749829498,
            "xValue": [
              "2013-06-01",
            ],
          },
          "selected": false,
          "y": 3213000.749829498,
        },
        {
          "custom": {
            "rawValue": 3314632.0388838113,
            "xValue": [
              "2013-07-01",
            ],
          },
          "selected": false,
          "y": 3314632.0388838113,
        },
        {
          "custom": {
            "rawValue": 3504006.725372224,
            "xValue": [
              "2013-08-01",
            ],
          },
          "selected": false,
          "y": 3504006.725372224,
        },
        {
          "custom": {
            "rawValue": 3819677.0195216006,
            "xValue": [
              "2013-09-01",
            ],
          },
          "selected": false,
          "y": 3819677.0195216006,
        },
        {
          "custom": {
            "rawValue": 3984064.438412456,
            "xValue": [
              "2013-10-01",
            ],
          },
          "selected": false,
          "y": 3984064.438412456,
        },
        {
          "custom": {
            "rawValue": 3403293.012026755,
            "xValue": [
              "2013-11-01",
            ],
          },
          "selected": false,
          "y": 3403293.012026755,
        },
        {
          "custom": {
            "rawValue": 1735405.3318556482,
            "xValue": [
              "2013-12-01",
            ],
          },
          "selected": false,
          "y": 1735405.3318556482,
        },
      ],
      "dataLabels": {
        "formatter": [Function],
      },
      "lineWidth": 3,
      "marker": {
        "enabled": false,
        "fillColor": null,
        "fillOpacity": 0.3,
        "lineColor": null,
        "lineWidth": 2,
        "radius": 4,
        "states": {
          "hover": {
            "fillColor": "#FFFFFF",
            "lineWidth": 3,
            "radius": 6,
          },
        },
        "symbol": "circle",
      },
      "name": "$trend_Cost",
      "showInLegend": false,
      "showInNavigator": true,
      "stickyTracking": false,
      "turboThreshold": 0,
      "yAxis": 0,
    },
  ],
  "title": {
    "text": null,
  },
  "tooltip": {
    "animation": false,
    "backgroundColor": "#FFFFFF",
    "borderColor": "#CCCCCC",
    "borderRadius": 10,
    "borderWidth": 1,
    "formatter": [Function],
    "style": {
      "fontFamily": "Open Sans",
    },
    "useHTML": true,
  },
  "xAxis": [
    {
      "categories": [
        "2009-11",
        "2009-12",
        "2010-01",
        "2010-02",
        "2010-03",
        "2010-04",
        "2010-05",
        "2010-06",
        "2010-07",
        "2010-08",
        "2010-09",
        "2010-10",
        "2010-11",
        "2010-12",
        "2011-01",
        "2011-02",
        "2011-03",
        "2011-04",
        "2011-05",
        "2011-06",
        "2011-07",
        "2011-08",
        "2011-09",
        "2011-10",
        "2011-11",
        "2011-12",
        "2012-01",
        "2012-02",
        "2012-03",
        "2012-04",
        "2012-05",
        "2012-06",
        "2012-07",
        "2012-08",
        "2012-09",
        "2012-10",
        "2012-11",
        "2012-12",
        "2013-01",
        "2013-02",
        "2013-03",
        "2013-04",
        "2013-05",
        "2013-06",
        "2013-07",
        "2013-08",
        "2013-09",
        "2013-10",
        "2013-11",
        "2013-12",
      ],
      "endOnTick": false,
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 0,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "overflow": "none",
        "style": {
          "color": "#5B6372",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "lineWidth": 1,
      "max": null,
      "min": null,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineWidth": 0,
      "minorTickColor": "#d1d1d7",
      "plotBands": [],
      "plotLines": [],
      "startOnTick": false,
      "tickColor": "#d1d1d7",
      "tickInterval": null,
      "tickWidth": 0,
      "tickmarkPlacement": "on",
      "title": {
        "enabled": false,
        "margin": 25,
        "style": {
          "color": "#5B6372",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
        "text": "X Axis title",
      },
      "type": "linear",
    },
  ],
  "yAxis": [
    {
      "endOnTick": true,
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 1,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "style": {
          "color": "#5B6372",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "lineWidth": 1,
      "max": 4844611.66696444,
      "min": 189976.80757868662,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineDashStyle": "Dot",
      "minorGridLineWidth": 0,
      "minorTickColor": "#d1d1d7",
      "minorTickWidth": 0,
      "plotBands": undefined,
      "showLastLabel": true,
      "stackLabels": {
        "allowOverlap": false,
        "crop": true,
        "enabled": false,
        "labelrank": 99999,
        "rotation": 0,
        "style": {
          "color": "#56535b",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "bold",
          "pointerEvents": "none",
          "textOutline": "none",
          "textOverflow": "ellipsis",
        },
      },
      "startOnTick": true,
      "tickColor": "#d1d1d7",
      "tickInterval": null,
      "tickWidth": 0,
      "tickmarkPlacement": "on",
      "title": {
        "enabled": false,
        "margin": 25,
        "style": {
          "color": "#5B6372",
          "fontFamily": "Open Sans",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
        "text": "Y Axis title",
      },
      "type": "linear",
    },
  ],
}
`;
